[Design] Change Tera to work similarly to mainline

DayKev opened the pull request:
> [!IMPORTANT]
> **This PR updates locales!**

> [!TIP]
> ### Follow Up Issues
>  <details><summary>Show</summary>
> <p>
> 
> - #889 
> - #986
> - #987 
> - #988
>   - requires #885
> - #989 
> - #997 
> 
> </p>
> </details> 
> 

## What are the changes the user will see?
- The player will be able to Terastallize their Pokémon in [the same way they can in Scarlet and Violet](https://bulbapedia.bulbagarden.net/wiki/Terastal_phenomenon#Effects) once they get a Tera Orb, with the effects also now matching mainline.
- Players can choose the Tera type of their starters based on that starter's existing types.
- ~Tera shards will be able to change the tera type of a Pokémon.~ To be added in a future PR after items have been refactored.

## Why am I making these changes?
Align Tera behavior with mainline.

## What are the changes from a developer perspective?
### Features
- Tera type of starters can be cycled on the starter select screen for dual-type Pokémon.
- Terastallization is enabled in battle when the player has a Tera Orb.
- Terastallization can be initiated from the fight UI by using the button to the left of the "Fight" and "Pokemon" commands.
- Terastallization lasts until the arena is reset or the Terastallized Pokémon faints.
- Only one Pokémon can be Terastallized at a time, and cannot be used on Mega or GMax Pokémon.
- Once Terastallization is used, the player cannot use it again until the Tera Orb is recharged, which happens in the `PartyHealPhase` that happens after every 10th wave. Alternatively, if the player has a Terapagos in their party it will automatically be recharged every wave. It will also automatically recharge on the E4 waves (181-190).

### Bugs Fixed
- Tera Stellar no longer modifies a Pokémon's defensive typing.
- Tera Stellar now properly boosts damage based on mainline behavior.
- Fixes #570 
- Fixes #577 
- Fixes #578 

### Misc Changes
- New test helper function added to Terastallize a Pokémon to any type.
- Terapagos locked to Stellar Tera type.
- Ogerpon locked to the Tera type of its matching form.
- Shedinja locked to Bug Tera type (custom).
- Embody Aspect is no longer partial.

## Screenshots/Videos
<details><summary>Tera button on fight UI</summary>
<p>

![image](https://github.com/user-attachments/assets/aeecd98e-606d-49ec-afe2-cfd0df4f5179)

</p>
</details> 
<details><summary>Terastallized Pokémon</summary>
<p>

![image](https://github.com/user-attachments/assets/5f523d9e-6eef-4e2c-9a1d-c1f0d966b548)

</p>
</details> 
<details><summary>Cursor replaced when about to Terastallize</summary>
<p>

![image](https://github.com/user-attachments/assets/926f9bd8-e752-4b1f-93e3-01b590a3819d)

</p>
</details> 
<details><summary>Light theme version of Tera cursor</summary>
<p>

![image](https://github.com/user-attachments/assets/40109170-6e9f-47f6-ab3c-98258100d9e4)

</p>
</details> 
<details><summary>Tera type shown on starter select screen</summary>
<p>

![image](https://github.com/user-attachments/assets/2510023d-79e4-49fc-bdf5-f2e6094061b5)

</p>
</details> 
<details><summary>Tera type shown on Pokémon summary screen</summary>
<p>

![image](https://github.com/user-attachments/assets/08f6d6ed-d83a-4362-9abc-2786bf619177)

</p>
</details> 

## How to test the changes?
Start a new Classic run to check that the player cannot Tera their Pokémon without a Tera Orb. Then use the following override and start a new run to test Tera behavior:
```ts
const overrides = {
  STARTING_MODIFIER_OVERRIDE: [{name: "TERA_ORB"}]
} satisfies Partial<InstanceType<typeof DefaultOverrides>>;
```

## Checklist
- [x] Otherwise: **I'm using `beta` as my base branch**
- [x] There is no overlap with another PR?
- [x] The PR is self-contained and cannot be split into smaller PRs?
- [x] Have I provided a clear explanation of the changes?
- [x] Have I tested the changes manually?
- [x] Are all unit tests still passing? (`npm run test:silent`)
  - [x] Have I created new automated tests (`npm run test:create`) or updated existing tests related to the PR's changes?
- [x] Have I provided screenshots/videos of the changes (if applicable)?
  - [x] Have I made sure that any UI change works for both UI themes (dark and light)?

#### Are there any localization additions or changes? If so:

- [x] Has a locales PR been created on the [locales](https://github.com/despair-games/poketernity-locales) repo?
  - [x] If so, please leave a link to it here: https://github.com/Despair-Games/poketernity-locales/pull/43
- [x] Have I added the `Localization` tag to this PR?

DayKev committed:
Replace tera code in `pokemon.ts`

DayKev committed:
Remove tera code from `battle-scene.ts`

DayKev committed:
Replace `.isTerastallized()` with `.terastallized`

DayKev committed:
Replace `.getTeraType()` with `.teraType`

DayKev committed:
Remove check for tera type being "unknown" in `move.ts`

DayKev committed:
Update `pokemon-forms.ts`

DayKev committed:
Remove Tera Shards

DayKev committed:
Update `run-info-ui-handler.ts`

DayKev committed:
Update Libero/Protean tests and combine them into one file

DayKev committed:
Update move effectiveness tests and add test helper to tera pokemon

DayKev committed:
Update tests: Freeze Dry, Tar Shot, Tera Blast, Tera Starstorm

DayKev committed:
Temporarily remove tera code from "The Expert Pokemon Breeder" ME

DayKev committed:
Temporarily disable tera-related code in trainer configs

DayKev committed:
Fix Dig, Dive, and Fly tests

DayKev committed:
Replace `!== ElementalType.UNKNOWN` checks with `.terastallized`

DayKev committed:
Fix "Damage Calculation" test

DayKev committed:
Rename `Pokemon.terastallized` to `Pokemon.isTerastallized`

DayKev committed:
Port the changes from PKR's tera refactor PR

Co-Authored-By: Xavion3 <xavion333@gmail.com>

DayKev committed:
Add Tera type choice to starter select

Co-Authored-By: Xavion3 <xavion333@gmail.com>

DayKev committed:
Change fight cursor to Tera icon when choosing to Terastallize

Co-Authored-By: Xavion3 <xavion333@gmail.com>

DayKev committed:
Add Tera type icon on summary screen

Co-Authored-By: Xavion3 <xavion333@gmail.com>

DayKev committed:
Fix text size of Tera cycle label in starter select

DayKev committed:
Prevent Mega and GMax pokemon from Terastallizing

DayKev committed:
Hide Tera icon and prevent cycling Tera type with IV graph displayed

Co-authored-by: Wlowscha <54003515+Wlowscha@users.noreply.github.com>

DayKev committed:
Fix imports in `phase-id.ts`

DayKev commented:
Tera phase timing will be incorrect until after #885 is merged.

DayKev committed:
Run prettier + eslint, add typedoc output to `.prettierignore`

DayKev committed:
Resolve `TODO` in `encounter-phase-utils.ts`

DayKev committed:
Resolve merge issues

DayKev committed:
Misc cleanup

DayKev committed:
Fix *ate abilities not changing type of Tera Blast/Starstorm

DayKev committed:
Embody Aspect is no longer partial

DayKev commented on src/data/init/init-abilities.ts:
```diff
@@ -202,6 +203,12 @@ import { TerrainType } from "#enums/terrain-type";
 import i18next from "i18next";
 
+/** Used for Aerialate, Refrigerate, Pixilate, Galvanize */
+const basicMoveTypeChangeAbilityCondition: PokemonAttackCondition = (user, _target, move) =>
```
A better name for this is welcome.

flx-sta replied:
So they all alter normal-type moves to something else.
- https://bulbapedia.bulbagarden.net/wiki/Galvanize_(Ability)
- https://bulbapedia.bulbagarden.net/wiki/Aerilate_(Ability)
- https://bulbapedia.bulbagarden.net/wiki/Pixilate_(Ability)
- https://bulbapedia.bulbagarden.net/wiki/Refrigerate_(Ability)

maybe:
```suggestion
const alterNormalTypeMoveCondition: PokemonAttackCondition = (user, _target, move) =>
```

or
```suggestion
const normalTypeConversionCondition: PokemonAttackCondition = (user, _target, move) =>
```

But I might be missing some context

Tempo-anon replied:
I've heard these abilities referred to as -ate abilities online but I don't know if that would be a good name. Also for reference, the only abilities that change move types are Aerilate/Galvanize/Pixilate/Refrigerate which change normal->flying/electric/fairy/ice and give a 20% boost, normalize (change everything to normal and give a 20% boost), and liquid voice (changes sound moves to water with no atk boost)

flx-sta replied:
https://pokemondb.net/pokebase/244961/what-are-all-of-the-ate-abilitites

flx-sta commented:
> ## TODO
> 
> - Replace placeholder assets copied from Xavion's PR (can be done later)
> - Fix Terastallization timing (waiting for #885, will probably be done in a follow-up)
> - Trainers (will be done in a follow-up)
> - Tera shards to change a Pokémon's tera type (pending items refactor)
> - Terraform Zero ability (will be done in a follow-up)

@DayKev are there already follow-up issues/tasks in place so it doesn't get lost?

flx-sta commented on src/battle-scene.ts:
```diff
@@ -580,8 +581,8 @@ export default class BattleScene extends SceneBase {
 
-    this.spriteSparkleHandler = new PokemonSpriteSparkleHandler();
-    this.spriteSparkleHandler.setup();
+    this.spriteTeraSparkleHandler = new PokemonSpriteTeraSparkleHandler();
+    this.spriteTeraSparkleHandler.setup();
```
NIT/Q: Why not call `setup()` in the contructor?

DayKev replied:
Dunno, I didn't write the class, I only renamed it.

flx-sta commented on src/battle-scene.ts:
```diff
@@ -1373,7 +1374,16 @@ export default class BattleScene extends SceneBase {
 
         for (const pokemon of this.getPlayerParty()) {
           pokemon.resetWaveData();
+          pokemon.resetTera(); // TODO: put this in resetBattleData?
```
Q: Is this `TODO` still relevant/open?

DayKev replied:
Yes, though it depends on whether we go through with having the arena reset every wave or not.

flx-sta replied:
Won't be able to resolve it in this PR I'd say

flx-sta commented on src/battle-scene.ts:
```diff
@@ -1373,7 +1374,16 @@ export default class BattleScene extends SceneBase {
+          pokemon.resetTera(); // TODO: put this in resetBattleData?
+
           applyAbAttrs<PostBattleInitAbAttr>(AbAttrFlag.POST_BATTLE_INIT, pokemon, false);
+
+          if (
+            pokemon.species.speciesId === SpeciesId.TERAPAGOS
+            || (this.gameMode.isClassic && isBetween(this.currentBattle.waveIndex, 181, 190))
+          ) {
+            this.arena.playerTerasUsed = 0;
+          }
```
I assume this is this:
> Upon capturing Terapagos at the end of [The Indigo Disk](https://bulbapedia.bulbagarden.net/wiki/The_Indigo_Disk), the Tera Orb will be infused with Terapagos' energy and will no longer require charging.

We should document that in a TLDR way

```suggestion

          // Upon capturing TERAPAGOS [...],  the Tera Orb will be infused with Terapagos' energy and will no longer require charging
          if (
            pokemon.species.speciesId === SpeciesId.TERAPAGOS
            || (this.gameMode.isClassic && isBetween(this.currentBattle.waveIndex, 181, 190))
          ) {
            this.arena.playerTerasUsed = 0;
          }
```

Although it seems that here the player hasn't caught it so not sure

Tempo-anon replied:
Would also add on to that comment about how it's automatically recharged from 181-190

flx-sta commented on src/data/init/init-abilities.ts:
```diff
@@ -202,6 +203,12 @@ import { TerrainType } from "#enums/terrain-type";
+const basicMoveTypeChangeAbilityCondition: PokemonAttackCondition = (user, _target, move) =>
+  move?.type === ElementalType.NORMAL
+  && (!move.hasAttr(VariableMoveTypeAttr)
+    || ([MoveId.TERA_BLAST, MoveId.TERA_STARSTORM].includes(move.id) && !user?.isTerastallized));
```
NIT: Please create a (frozen) array for those. e.g. in `move-utils`

```ts
const teraMoveIds = Object.freeze([MoveId.TERA_BLAST, MoveId.TERA_SANDSTORM]);
```

and then
```suggestion
    || (teraMoveIds.includes(move.id) && !user?.isTerastallized));
```

----

ALT: Create a helper method `move.isTeraMove()` (it could still check that array)

flx-sta commented on src/data/init/init-abilities.ts:
```diff
@@ -202,6 +203,12 @@ import { TerrainType } from "#enums/terrain-type";
+  && (!move.hasAttr(VariableMoveTypeAttr)
+    || ([MoveId.TERA_BLAST, MoveId.TERA_STARSTORM].includes(move.id) && !user?.isTerastallized));
+
 // prettier-ignore
```
Q: is the prettier-ignore here still relevant?

DayKev replied:
It was added to prevent prettier from mangling the move/ability arrays.

flx-sta commented on src/data/init/init-moves.ts:
```diff
@@ -3259,7 +3259,7 @@ export function initMoves() {
       .attr(TeraBlastPowerAttr)
       .attr(StatStageChangeAttr, [Stat.ATK, Stat.SPATK], -1, true, {
-        condition: (user, _target, _move) => user.isTerastallized() && user.isOfType(ElementalType.STELLAR),
+        condition: (user, _target, _move) => user.isTerastallized && user.isOfType(ElementalType.STELLAR),
```
NIT: If you dont' need params after the optionals you can omit them:
```suggestion
        condition: (user) => user.isTerastallized && user.isOfType(ElementalType.STELLAR),
```

flx-sta commented on src/data/init/init-moves.ts:
```diff
@@ -3476,7 +3476,7 @@ export function initMoves() {
       .attr(TeraMoveCategoryAttr)
       .attr(TeraStarstormTypeAttr)
       .attr(VariableTargetAttr, (user, _target, _move) =>
-        user.species.speciesId === SpeciesId.TERAPAGOS && user.isTerastallized()
+        user.species.speciesId === SpeciesId.TERAPAGOS && user.isTerastallized
```
NIT: If you dont' need params after the optionals you can omit them:
```suggestion
     .attr(VariableTargetAttr, (user) =>
        user.species.speciesId === SpeciesId.TERAPAGOS && user.isTerastallized
```

flx-sta commented on src/data/moves/move-attrs/add-type-attr.ts:
```diff
@@ -39,6 +39,6 @@ export class AddTypeAttr extends MoveEffectAttr {
 
   override getCondition(): MoveConditionFunc {
-    return (_user, target, _move) => !target.isTerastallized() && !target.getTypes().includes(this.type);
+    return (_user, target, _move) => !target.isTerastallized && !target.getTypes().includes(this.type);
```
NIT: If you dont' need params after the optionals you can omit them:

```suggestion
    return (_user, target) => !target.isTerastallized && !target.getTypes().includes(this.type);
```

flx-sta commented on src/data/mystery-encounters/utils/encounter-phase-utils.ts:
```diff
@@ -117,6 +118,7 @@ export interface EnemyPokemonConfig {
   modifierConfigs?: HeldModifierConfig[];
   tags?: BattlerTagType[];
   dataSource?: PokemonData;
+  tera?: ElementalType;
```
Q: Why `tera` here and `teraType` on Pokemon?

DayKev replied:
Missed renaming, will change.

flx-sta commented on src/data/pokemon-forms.ts:
```diff
@@ -194,9 +176,7 @@ export class SpeciesFormChangeTeraTrigger extends SpeciesFormChangeTrigger {
  * @extends SpeciesFormChangeTrigger
  */
 export class SpeciesFormChangeLapseTeraTrigger extends SpeciesFormChangeTrigger {
```
Q: So what's the purpose of this class now?

flx-sta commented on src/data/pokemon-forms.ts:
```diff
@@ -168,24 +167,7 @@ export class SpeciesDefaultFormMatchTrigger extends SpeciesFormChangeTrigger {
  * @extends SpeciesFormChangeTrigger
  */
 export class SpeciesFormChangeTeraTrigger extends SpeciesFormChangeTrigger {
```
Q: So what's the purpose of this class now?

DayKev replied:
It's a manual form change trigger. It's separate so that it can be triggered separately.

flx-sta commented on src/data/trainer-config.ts:
```diff
@@ -250,11 +250,41 @@ export const trainerPartyTemplates = {
+  public teraMode: TeraAIMode;
+  public instantTeras: number[] = [];
+
+  /**
+   * @param canTerastallize Whether this trainer is allowed to tera
+   */
```
NIT: this doesn't match with the constructor params

flx-sta commented on src/data/trainer-config.ts:
```diff
@@ -640,6 +672,36 @@ export class TrainerConfig {
   }
 
+  /**
+   * Sets random pokemon from the trainers team to instant tera. Uses their specialty types is they have one.
```
Typo
```suggestion
   * Sets random pokemon from the trainers team to instant tera. Uses their specialty types if they have one.
```

flx-sta commented on src/data/trainer-config.ts:
```diff
@@ -640,6 +672,36 @@ export class TrainerConfig {
+   * @param count The amount of pokemon to have instant tera
+   * @returns this
+   */
+  setRandomTeraModifiers(count: () => number): TrainerConfig {
```
Q: Why is the param called `count` but it's a fn returning a number? That's so odd and the name might be misleading then

DayKev replied:
I have changes planned for this in a follow-up so I'll address it then too.

flx-sta commented on src/data/trainer-config.ts:
```diff
@@ -640,6 +672,36 @@ export class TrainerConfig {
+   * @returns this
+   */
+  setRandomTeraModifiers(count: () => number): TrainerConfig {
+    this.genAIFuncs.push((party: EnemyPokemon[]) => {
+      const partyMemberIndexes = new Array(party.length).fill(null).map((_, i) => i);
+      for (let t = 0; t < Math.min(count(), party.length); t++) {
+        const randomIndex = randSeedItem(partyMemberIndexes);
+        partyMemberIndexes.splice(partyMemberIndexes.indexOf(randomIndex), 1);
+        if (this.specialtyTypes?.length) {
+          party[randomIndex].teraType = randSeedItem(this.specialtyTypes);
+        }
+        this.trainerAI.setInstantTera(randomIndex);
+      }
+    });
```
NIT:
```suggestion
    this.genAIFuncs.push((party: EnemyPokemon[]) => {
      const { length } = party;
      const partyMemberIndexes = Array.from({ length }).map((_, i) => i);
      for (let t = 0; t < Math.min(count(), length); t++) {
        const randomIndex = randSeedItem(partyMemberIndexes);
        partyMemberIndexes.splice(partyMemberIndexes.indexOf(randomIndex), 1);
        if (this.specialtyTypes?.length) {
          party[randomIndex].teraType = randSeedItem(this.specialtyTypes);
        }
        this.trainerAI.setInstantTera(randomIndex);
      }
    });
```

flx-sta commented on src/data/trainer-config.ts:
```diff
@@ -640,6 +672,36 @@ export class TrainerConfig {
+  /**
+   * Sets random pokemon from the trainers team to instant tera. Uses their specialty types is they have one.
+   * @param count The amount of pokemon to have instant tera
+   * @returns this
```
NIT: Match existing docs:
```suggestion
   * @returns this {@linkcode TrainerConfig}
```

flx-sta commented on src/data/trainer-config.ts:
```diff
@@ -640,6 +672,36 @@ export class TrainerConfig {
+  /**
+   * Sets a specific pokemon to instant tera
+   * @param index The index within the team to have instant tera
+   * @returns this
```
NIT: Match existing docs:
```suggestion
   * @returns this {@linkcode TrainerConfig}
```

flx-sta commented on src/data/trainer-config.ts:
```diff
@@ -1648,21 +1712,21 @@ export function getSpeciesFilterRandomPartyMemberFunc(
-  );
-  return ret;
-}
+// function getSpecificTeraModifier(
+//   party: EnemyPokemon[],
+//   partySlot: number,
+//   teraType: ElementalType,
+// ): PersistentModifier[] {
+//   const ret: PersistentModifier[] = [];
+//   ret.push(
+//     modifierTypes
+//       .TERA_SHARD()
+//       .generateType([], [teraType])!
+//       .withIdFromFunc(modifierTypes.TERA_SHARD)
+//       .newModifier(party[partySlot]) as PersistentModifier,
+//   );
+//   return ret;
+// }
```
Q: forgot to remove it? if not add a note/todo why it's not

flx-sta commented on src/data/trainer-config.ts:
```diff
@@ -1308,9 +1371,10 @@ export class TrainerConfig {
-    this.setGenModifiersFunc((party) =>
-      getRandomTeraModifiers(party, 1, specialtyTypes.length ? specialtyTypes : undefined),
-    );
+    // this.setGenModifiersFunc((party) =>
+    //   getRandomTeraModifiers(party, 1, specialtyTypes.length ? specialtyTypes : undefined),
+    // );
```
Q: Forgot to remove it? If not add a note/todo why it's not

flx-sta commented on src/data/trainer-config.ts:
```diff
@@ -1308,9 +1371,10 @@ export class TrainerConfig {
+    // this.setGenModifiersFunc((party) =>
+    //   getRandomTeraModifiers(party, 1, specialtyTypes.length ? specialtyTypes : undefined),
+    // );
+    this.setRandomTeraModifiers(() => 1); // TODO: is this what we want?
```
Q: This Todo still relevant/open?

Tempo-anon replied:
Bleh I should have documented this better when I changed it. This gives all E4 a tera of their type to a random Pokemon on the team

flx-sta commented on src/data/trainer-config.ts:
```diff
@@ -1671,19 +1735,19 @@ function getSpecificTeraModifier(
-  }
-  return ret;
-}
+// function getRandomTeraModifiers(party: EnemyPokemon[], count: number, types?: ElementalType[]): PersistentModifier[] {
+//   const ret: PersistentModifier[] = [];
+//   const partyMemberIndexes = new Array(party.length).fill(null).map((_, i) => i);
+//   for (let t = 0; t < Math.min(count, party.length); t++) {
+//     const randomIndex = randSeedItem(partyMemberIndexes);
+//     partyMemberIndexes.splice(partyMemberIndexes.indexOf(randomIndex), 1);
+//     ret.push(
+//       modifierTypes
+//         .TERA_SHARD()
+//         .generateType([], [randSeedItem(types ? types : party[randomIndex].getTypes())])!
+//         .withIdFromFunc(modifierTypes.TERA_SHARD)
+//         .newModifier(party[randomIndex]) as PersistentModifier,
+//     ); // TODO: is the bang correct?
+//   }
+//   return ret;
+// }
```
Q: forgot to remove it? if not add a note/todo why it's not

flx-sta commented on src/data/trainer-configs/champion-trainer-configs.ts:
```diff
@@ -232,17 +229,17 @@ export const championTrainerConfigs: TrainerConfigs = {
-      ]; //TODO: is the bang correct?
-    }),
+    ),
+  // .setGenModifiersFunc((party) => {
+  //   const teraPokemon = party[4];
+  //   return [
+  //     modifierTypes
+  //       .TERA_SHARD()
+  //       .generateType([], [ElementalType.FIRE])!
+  //       .withIdFromFunc(modifierTypes.TERA_SHARD)
+  //       .newModifier(teraPokemon) as PersistentModifier,
+  //   ]; //TODO: is the bang correct?
+  // }),
```
Q: forgot to remove it? if not add a note/todo why it's not

flx-sta commented on src/data/trainer-configs/champion-trainer-configs.ts:
```diff
@@ -321,28 +318,28 @@ export const championTrainerConfigs: TrainerConfigs = {
-      ]; //TODO: is the bang correct?
-    }),
+    ),
+  // .setGenModifiersFunc((party) => {
+  //   const teraPokemon = party[4];
+  //   let teraType: ElementalType;
+  //   switch (teraPokemon.species.speciesId) {
+  //     case SpeciesId.DECIDUEYE:
+  //       teraType = ElementalType.GHOST;
+  //       break;
+  //     case SpeciesId.INCINEROAR:
+  //       teraType = ElementalType.DARK;
+  //       break;
+  //     default:
+  //       teraType = ElementalType.WATER;
+  //   }
+  //   return [
+  //     modifierTypes
+  //       .TERA_SHARD()
+  //       .generateType([], [teraType])!
+  //       .withIdFromFunc(modifierTypes.TERA_SHARD)
+  //       .newModifier(teraPokemon) as PersistentModifier,
+  //   ]; //TODO: is the bang correct?
+  // }),
```
Q: forgot to remove it? if not add a note/todo why it's not

flx-sta commented on src/data/trainer-configs/champion-trainer-configs.ts:
```diff
@@ -386,17 +383,17 @@ export const championTrainerConfigs: TrainerConfigs = {
-      ]; //TODO: is the bang correct?
-    }),
+    ),
+  // .setGenModifiersFunc((party) => {
+  //   const teraPokemon = party[4];
+  //   return [
+  //     modifierTypes
+  //       .TERA_SHARD()
+  //       .generateType([], [ElementalType.FLYING])!
+  //       .withIdFromFunc(modifierTypes.TERA_SHARD)
+  //       .newModifier(teraPokemon) as PersistentModifier,
+  //   ]; //TODO: is the bang correct?
+  // }),
```
Q: forgot to remove it? if not add a note/todo why it's not

flx-sta commented on src/data/trainer-configs/champion-trainer-configs.ts:
```diff
@@ -429,28 +426,28 @@ export const championTrainerConfigs: TrainerConfigs = {
-      ]; //TODO: is the bang correct?
-    }),
+    ),
+  // .setGenModifiersFunc((party) => {
+  //   const teraPokemon = party[4];
+  //   let teraType: ElementalType;
+  //   switch (teraPokemon.species.speciesId) {
+  //     case SpeciesId.MEOWSCARADA:
+  //       teraType = ElementalType.GRASS;
+  //       break;
+  //     case SpeciesId.SKELEDIRGE:
+  //       teraType = ElementalType.FIRE;
+  //       break;
+  //     default:
+  //       teraType = ElementalType.WATER;
+  //   }
+  //   return [
+  //     modifierTypes
+  //       .TERA_SHARD()
+  //       .generateType([], [teraType])!
+  //       .withIdFromFunc(modifierTypes.TERA_SHARD)
+  //       .newModifier(teraPokemon) as PersistentModifier,
+  //   ]; //TODO: is the bang correct?
+  // }),
```
Q: forgot to remove it? if not add a note/todo why it's not

flx-sta commented on src/data/trainer-configs/evil-boss-trainer-configs.ts:
```diff
@@ -657,17 +655,17 @@ export const evilBossTrainerConfigs: TrainerConfigs = {
-      ]; //TODO: is the bang correct?
-    }),
+    ),
+  // .setGenModifiersFunc((party) => {
+  //   const teraPokemon = party[4];
+  //   return [
+  //     modifierTypes
+  //       .TERA_SHARD()
+  //       .generateType([], [teraPokemon.species.type1])!
+  //       .withIdFromFunc(modifierTypes.TERA_SHARD)
+  //       .newModifier(teraPokemon) as PersistentModifier,
+  //   ]; //TODO: is the bang correct?
+  // }),
```
Q: forgot to remove it? if not add a note/todo why it's not

flx-sta commented on src/data/trainer-configs/evil-boss-trainer-configs.ts:
```diff
@@ -725,15 +723,15 @@ export const evilBossTrainerConfigs: TrainerConfigs = {
-      ]; //TODO: is the bang correct?
-    }),
+    ),
+  // .setGenModifiersFunc((party) => {
+  //   const teraPokemon = party[0];
+  //   return [
+  //     modifierTypes
+  //       .TERA_SHARD()
+  //       .generateType([], [teraPokemon.species.type1])!
+  //       .withIdFromFunc(modifierTypes.TERA_SHARD)
+  //       .newModifier(teraPokemon) as PersistentModifier,
+  //   ]; //TODO: is the bang correct?
+  // }),
```
Q: forgot to remove it? if not add a note/todo why it's not

flx-sta commented on src/data/trainer-configs/rival-trainer-configs.ts:
```diff
@@ -315,17 +314,17 @@ export const rivalTrainerConfigs: TrainerConfigs = {
-      ]; // TODO: is the bang correct?
-    }),
+    .setSpeciesFilter((species) => species.baseTotal >= 540),
+  // .setGenModifiersFunc((party) => {
+  //   const starter = party[0];
+  //   return [
+  //     modifierTypes
+  //       .TERA_SHARD()
+  //       .generateType([], [starter.species.type1])!
+  //       .withIdFromFunc(modifierTypes.TERA_SHARD)
+  //       .newModifier(starter) as PersistentModifier,
+  //   ]; // TODO: is the bang correct?
+  // }),
```
Q: forgot to remove it? if not add a note/todo why it's not

flx-sta commented on src/data/trainer-configs/rival-trainer-configs.ts:
```diff
@@ -413,17 +412,17 @@ export const rivalTrainerConfigs: TrainerConfigs = {
-      ]; //TODO: is the bang correct?
-    }),
+    ),
+  // .setGenModifiersFunc((party) => {
+  //   const starter = party[0];
+  //   return [
+  //     modifierTypes
+  //       .TERA_SHARD()
+  //       .generateType([], [starter.species.type1])!
+  //       .withIdFromFunc(modifierTypes.TERA_SHARD)
+  //       .newModifier(starter) as PersistentModifier,
+  //   ]; //TODO: is the bang correct?
+  // }),
```
Q: forgot to remove it? if not add a note/todo why it's not

flx-sta commented on src/data/trainer-configs/rival-trainer-configs.ts:
```diff
@@ -519,15 +518,15 @@ export const rivalTrainerConfigs: TrainerConfigs = {
-      ]; // TODO: is the bang correct?
-    }),
+    ),
+  // .setGenModifiersFunc((party) => {
+  //   const starter = party[0];
+  //   return [
+  //     modifierTypes
+  //       .TERA_SHARD()
+  //       .generateType([], [starter.species.type1])!
+  //       .withIdFromFunc(modifierTypes.TERA_SHARD)
+  //       .newModifier(starter) as PersistentModifier,
+  //   ]; // TODO: is the bang correct?
+  // }),
```
Q: forgot to remove it? if not add a note/todo why it's not

flx-sta commented on src/enums/common-anim.ts:
```diff
@@ -48,4 +48,5 @@ export enum CommonAnim {
   GRASSY_TERRAIN,
   PSYCHIC_TERRAIN,
   LOCK_ON = 2120,
+  TERASTALLIZE,
```
Q: Should this follow after the artificially high `LOCK_ON = 2120`?

DayKev replied:
I don't think it matters? Not sure. Inner might know, he's been doing work on the animations system; or possibly cam.

DayKev replied:
Asked Inner, it doesn't matter.

flx-sta commented on src/enums/phase-id.ts:
Dang... that's a lot of detail work

flx-sta commented on src/field/pokemon.ts:
```diff
@@ -100,7 +100,7 @@ import { VariableMoveTypeMultiplierAttr } from "#app/data/moves/move-attrs/varia
 import { starterPassiveAbilities } from "#app/data/passives";
 import type { SpeciesEvolutionCondition, SpeciesFormEvolution } from "#app/data/pokemon-evolutions";
-import type { SpeciesFormChange } from "#app/data/pokemon-forms";
+import { SpeciesFormChangeLapseTeraTrigger, type SpeciesFormChange } from "#app/data/pokemon-forms";
```
```suggestion
import { SpeciesFormChangeLapseTeraTrigger, type SpeciesFormChange } from "#app/data/pokemon-forms";
import type { SpeciesFormChange } from "#app/data/pokemon-forms";
```

Or did I miss the mark this time :D

DayKev replied:
Mixed imports (`import { X, type Y } from Z`) are okay, it's only when all imports from the file are type imports (`import { type X, type Y } from Z`) that it should be brought outside the brackets. The new ESLint rule will catch this for us so we don't need to worry about it.

flx-sta commented on src/field/pokemon.ts:
```diff
@@ -269,6 +268,9 @@ export abstract class Pokemon extends Phaser.GameObjects.Container {
   public switchOutStatus: boolean;
   public evoCounter: number;
+  protected _teraType: ElementalType;
+  public isTerastallized: boolean = false; // TODO: put this in battle data?
+  public stellarTypesBoosted: ElementalType[] = []; // TODO: put this in battle data?
```
Q: Are those TODOs still open?

flx-sta commented on src/field/pokemon.ts:
```diff
@@ -410,6 +417,48 @@ export abstract class Pokemon extends Phaser.GameObjects.Container {
     }
   }
 
+  public get teraType(): ElementalType {
```
NIT: tsdoc

flx-sta commented on src/field/pokemon.ts:
```diff
@@ -410,6 +417,48 @@ export abstract class Pokemon extends Phaser.GameObjects.Container {
+    return this._teraType;
+  }
+
+  public set teraType(value: ElementalType) {
```
NIT: tsdoc

flx-sta replied:
@DayKev I assume you marked this as resolved as it's simply a plain getter for the private value?

DayKev replied:
The TSDocs are on the getter, which applies it to all uses.
![image](https://github.com/user-attachments/assets/ba563762-35f4-4035-b864-193ea781748c)

If I were to add separate TSDocs to the setter, they would be combined in the display rather than shown separately.

flx-sta replied:
Forgot about that. Thanks

flx-sta commented on src/field/pokemon.ts:
```diff
@@ -410,6 +417,48 @@ export abstract class Pokemon extends Phaser.GameObjects.Container {
+    this._teraType = value;
+  }
+
+  public resetTera(): void {
```
NIT: tsdoc

flx-sta commented on src/field/pokemon.ts:
```diff
@@ -1376,9 +1429,9 @@ export abstract class Pokemon extends Phaser.GameObjects.Container {
-      if (teraType !== ElementalType.UNKNOWN) {
+    if (includeTeraType && this.isTerastallized) {
+      const teraType = this.teraType;
+      if (!(forDefend && teraType === ElementalType.STELLAR)) {
```
Q: Does `ElementalType.UNKNOWN` not need any special handling anymore?

DayKev replied:
Checking for `UNKNOWN` was the old way of checking if a pokemon was Terastallized or not, because non-Terastallized pokemon would have a Tera type of `UNKNOWN` and only Terastallized pokemon would have a Tera type. Now, all pokemon have a Tera type by default.

flx-sta commented on src/field/pokemon.ts:
```diff
@@ -3061,16 +3094,28 @@ export abstract class Pokemon extends Phaser.GameObjects.Container {
     const stabMultiplier = new NumberHolder(1);
-    if (matchesSourceType) {
+
+    if (matchesSourceType && moveType !== ElementalType.STELLAR) {
       stabMultiplier.value += 0.5;
     }
+
     applyMoveAttrs(CombinedPledgeStabBoostAttr, source, this, move, stabMultiplier);
-    if (sourceTeraType !== ElementalType.UNKNOWN && sourceTeraType === moveType) {
-      stabMultiplier.value += 0.5;
+
+    if (source.isTerastallized) {
+      if (sourceTeraType === moveType && moveType !== ElementalType.STELLAR) {
+        stabMultiplier.value += 0.5;
+      }
+
+      if (
+        source.teraType === ElementalType.STELLAR
+        && (!source.stellarTypesBoosted.includes(moveType) || source.species.speciesId === SpeciesId.TERAPAGOS)
+      ) {
+        stabMultiplier.value += matchesSourceType ? 0.5 : 0.2;
+      }
     }
 
     applyAbFunc<StabBoostAbAttr>(AbAttrFlag.STAB_BOOST, source, simulated, stabMultiplier);
```
NIT: Might wanna move the whole `stab` logic into an enclosed helper method (out of scope possibly)

flx-sta commented on src/field/pokemon.ts:
```diff
@@ -3393,6 +3438,16 @@ export abstract class Pokemon extends Phaser.GameObjects.Container {
     return maxForms.includes(this.getFormKey());
   }
 
+  isMega(): boolean {
```
NIT: tsdoc

flx-sta commented on src/field/pokemon.ts:
```diff
@@ -3393,6 +3438,16 @@ export abstract class Pokemon extends Phaser.GameObjects.Container {
   }
 
+  isMega(): boolean {
+    const megaForms = [
+      SpeciesFormKey.MEGA,
+      SpeciesFormKey.MEGA_X,
+      SpeciesFormKey.MEGA_Y,
+      SpeciesFormKey.PRIMAL,
+    ] as string[];
+    return megaForms.includes(this.getFormKey());
```
Q: Can't you define the array as `string[]` to begin with?
```suggestion
    const megaForms: string[] = [
      SpeciesFormKey.MEGA,
      SpeciesFormKey.MEGA_X,
      SpeciesFormKey.MEGA_Y,
      SpeciesFormKey.PRIMAL,
    ];
    return megaForms.includes(this.getFormKey());
```

DayKev replied:
Probably, this was copy-pasted. Will check.

flx-sta commented on src/field/trainer.ts:
```diff
@@ -618,6 +620,12 @@ export default class Trainer extends Phaser.GameObjects.Container {
     return [];
   }
 
+  public genAI(party: EnemyPokemon[]): void {
```
NIT: tsdoc

flx-sta commented on src/field/trainer.ts:
```diff
@@ -739,4 +747,13 @@ export default class Trainer extends Phaser.GameObjects.Container {
     });
   }
+
+  public shouldTera(pokemon: EnemyPokemon): boolean {
```
NIT: tsdoc

flx-sta commented on src/ui/handlers/fight-ui-handler.ts:
```diff
@@ -18,8 +20,8 @@ import { MoveCategory } from "#enums/move-category";
-import { UiHandler } from "./abstract-ui-handler";
-import type { CommandUiHandler } from "./command-ui-handler";
+
+type FightCommand = BattleCommand.FIGHT | BattleCommand.TERA;
```
NIT: Probably should define that somewhere more global. I've seen that definition in a previous file before without this "helper" type being used

DayKev replied:
True, will do.

flx-sta commented on src/ui/handlers/run-info-ui-handler.ts:
```diff
@@ -487,17 +487,11 @@ export class RunInfoUiHandler extends UiHandler {
+    // Old behavior: Determining which Terastallize Modifier belongs to which Pokemon
     // Creates a dictionary {PokemonId: TeraShardType}
+
+    // TODO: get tera'd pokemon
```
Q: TODO still relevant/open?

DayKev replied:
Oh yeah, I'll make an issue for this too.

flx-sta commented on src/ui/handlers/run-info-ui-handler.ts:
```diff
@@ -511,13 +505,16 @@ export class RunInfoUiHandler extends UiHandler {
       const enemyIcon = globalScene.addPokemonIcon(enemy, 0, 0, 0, 0);
+
       // Applying Terastallizing Type tint to Pokemon icon
+      // TODO: update this
```
Q: TODO still relevant/open?

flx-sta reviewed (commented):
### Pure Code-Review
<img width="140" alt="image" src="https://github.com/user-attachments/assets/62e151e0-7ecb-4070-9069-6f3e595d8b1f" />

Tempo-anon commented on src/data/trainer-config.ts:
```diff
@@ -250,11 +250,41 @@ export const trainerPartyTemplates = {
 
+class TrainerAI {
+  public teraMode: TeraAIMode;
+  public instantTeras: number[] = [];
```
I remember there was some talk of limiting 1 tera per team. Will that also be true for enemy trainers?

flx-sta replied:
> 1 of each per Team

Which includes Tera

[<img width="468" alt="image" src="https://github.com/user-attachments/assets/471d4313-e44a-434a-af2c-92f3e47dfaa2" />](https://discord.com/channels/1312659961502761000/1343699671524048957/1347116929218908201)

DayKev replied:
Yes, though MEs might want to break the rule. I'll double check whether MEs also use this field or not.

Tempo-anon replied:
I don't think any MEs do. The only MEs that might have to do with tera are the expert breeder which has a steel clefable and mysterious challengers which calls a gym leader. There's also the clown event that changes types but that's unrelated to tera at all (maybe it should?)

DayKev replied:
This field is set by MEs (currently only expert breeder as mentioned) to determine which Pokemon to Terastallize. Since we might decide that a new ME in the future is allowed to break the rules and have more than one pokemon Terastallize, it'll need to stay an array for now.

Tempo-anon commented on src/field/pokemon.ts:
```diff
@@ -410,6 +417,48 @@ export abstract class Pokemon extends Phaser.GameObjects.Container {
+            return ElementalType.ROCK;
+        }
+
+      case SpeciesId.SHEDINJA:
+        return ElementalType.BUG;
```
Is this custom?

DayKev replied:
I think so. I'll add a note.

DayKev commented:
> > ## TODO
> > 
> > * Replace placeholder assets copied from Xavion's PR (can be done later)
> > * Fix Terastallization timing (waiting for [[Feature] Add uniform dynamic turn command scheduling #885](https://github.com/Despair-Games/poketernity/pull/885), will probably be done in a follow-up)
> > * Trainers (will be done in a follow-up)
> > * Tera shards to change a Pokémon's tera type (pending items refactor)
> > * Terraform Zero ability (will be done in a follow-up)
> 
> @DayKev are there already follow-up issues/tasks in place so it doesn't get lost?

Created issues for the ones that didn't already have them.

flx-sta commented:
> Created issues for the ones that didn't already have them.

I'mma link them here:
- #989 
- #988 
- #987 
- #986

DayKev committed:
Add TSDocs to `Pokemon#teraType`, `resetTera`, `isMega`

DayKev committed:
Add `TERA_MOVES` constant in `constants/move-constants.ts`

Rename `basicMoveTypeChangeAbilityCondition`
to `normalTypeMoveConversionCondition`

Add comment explaining the Tera Orb recharge behavior

DayKev committed:
Move `type FightCommand` to `/@types/`; update comment in `pokemon.ts`

DayKev committed:
Remove unused param in `add-type-attr.ts`

Rename `tera` to `teraType` on `encounter-phase-utils.ts`

Fix bug preventing Normal type tera in `encounter-phase-utils.ts`

DayKev committed:
Update TSDocs in `trainer-config.ts`, `trainer.ts`, `tera-ai-mode.ts`

Apply suggestion to `TrainerConfig#setRandomTeraModifiers`

DayKev committed:
Remove unused params in `init-moves.ts`

DayKev committed:
Remove leftover commented code in `trainer-config.ts`

DayKev committed:
Put `TODO`s in place

DayKev committed:
Add TSDocs to `Trainer#shouldTera`

DayKev committed:
Change which class the Tera form change classes extend for clarity

DayKev committed:
Add TSDoc to `Trainer#genAI`

DayKev committed:
Update comment in `battle-scene.ts` regarding Tera Orb recharging

DayKev committed:
Update `TODO` comment in `battle-scene.ts`

DayKev commented on src/data/init/init-moves.ts:
```diff
@@ -3259,7 +3259,7 @@ export function initMoves() {
       .attr(TeraBlastPowerAttr)
       .attr(StatStageChangeAttr, [Stat.ATK, Stat.SPATK], -1, true, {
-        condition: (user, _target, _move) => user.isTerastallized() && user.isOfType(ElementalType.STELLAR),
+        condition: (user) => user.isTerastallized && user.isOfType(ElementalType.STELLAR),
```
I'm not sure it's best to remove the unused params in cases like these since in the future if it's necessary to use them it's not obvious to developers that they're available to be used.

flx-sta replied:
Feel free to restore them then if you don't feel confident about that change:

```suggestion
        condition: (user, _target, _move) => user.isTerastallized && user.isOfType(ElementalType.STELLAR),
```

DayKev commented:
Updates based on reviews starting at `537801e`

flx-sta commented on src/battle-scene.ts:
```diff
@@ -1373,7 +1373,18 @@ export default class BattleScene extends SceneBase {
+            // In Scarlet/Violet, the player's Tera Orb automatically recharges after every battle once they've caught Terapagos
+            pokemon.species.speciesId === SpeciesId.TERAPAGOS
+            // The player's Tera Orb also automatically recharges when fighting the Elite 4 or in Area Zero (the endgame area)
+            || (this.gameMode.isClassic && this.currentBattle.waveIndex > 180)
```
NIT: Since this is the wave where apparently Elite-4 start, can we give this wave-index a name as a constant so it's not a "magic-number"?

flx-sta commented on src/constants/move-constants.ts:
```diff
@@ -0,0 +1,3 @@
+import { MoveId } from "#enums/move-id";
+
+export const TERA_MOVES = Object.freeze([MoveId.TERA_BLAST, MoveId.TERA_STARSTORM]);
```
NIT: @Tempo-anon recently asked about the const pattern, as it's inconsistent. It seems we have chosen `PascalCase` instead of `UPPER_SNAKE_CASE`.

DayKev replied:
For now I'm going to keep matching the current convention which is `ALL_CAPS`. If we decide to switch to `PascalCase` then all the constants can be converted then.

flx-sta replied:
I've started a [poll on Discord](https://discord.com/channels/1312659961502761000/1343699671524048957/1362470863433236782) to resolve this in the long run. For now let's just ignore it

flx-sta commented on src/data/moves/move-attrs/add-type-attr.ts:
```diff
@@ -39,6 +39,6 @@ export class AddTypeAttr extends MoveEffectAttr {
 
   override getCondition(): MoveConditionFunc {
-    return (_user, target, _move) => !target.isTerastallized() && !target.getTypes().includes(this.type);
+    return (_user, target) => !target.isTerastallized && !target.getTypes().includes(this.type);
```
Feel free to return them then if you don't feel confident about that change
```suggestion
    return (_user, target, _move) => !target.isTerastallized && !target.getTypes().includes(this.type);
```

flx-sta commented on src/enums/tera-ai-mode.ts:
```diff
@@ -0,0 +1,9 @@
+  /** The trainer will Tersastallize the configured pokemon on the first turn it's sent out */
+  INSTANT,
+  /** (Currently unused) The trainer will choose when to Terastallize their pokemon based on the current battle state */
+  SMART,
```
And here I thought there was some kind of hidden complex smart logic involved :D

flx-sta reviewed (dismissed):
Looks good

DayKev committed:
Add constant `ELITE_FOUR_START_WAVE`

Re-add unused params in anonymous functions

DayKev committed:
Update locales

flx-sta commented on src/data/special-waves.ts:
```diff
@@ -30,6 +30,7 @@ export const EVIL_ADMIN_2_WAVE = 114;
 export const EVIL_BOSS_1_WAVE = 115;
 export const EVIL_BOSS_2_WAVE = 165;
 
+export const ELITE_FOUR_START_WAVE = 181;
```
Could have just used `>= ELITE_FOUR_1_WAVE` then, lol.
Didn't know these are there

DayKev replied:
Oh you're right, for some reason I was thinking it wouldn't apply if you used Tera on wave 181 and then transitioned to 182 if I did it that way. I'll remove the unnecessary extra constant and use that one.

DayKev committed:
Remove unnecessary constant and use existing one

innerthunder commented on src/data/init/init-abilities.ts:
```diff
@@ -202,6 +204,11 @@ import { TerrainType } from "#enums/terrain-type";
 import { WeatherType } from "#enums/weather-type";
 import i18next from "i18next";
 
+/** Used for Aerialate, Refrigerate, Pixilate, Galvanize */
+const normalTypeMoveConversionCondition: PokemonAttackCondition = (user, _target, move) =>
+  move?.type === ElementalType.NORMAL
+  && (!move.hasAttr(VariableMoveTypeAttr) || (TERA_MOVES.includes(move.id) && !user?.isTerastallized));
```
This should be moved to `ability-utils.ts`

innerthunder commented on src/data/init/init-abilities.ts:
```diff
@@ -1622,29 +1629,25 @@ export function initAbilities() {
       .edgeCase(), // Does not inflict poison if user gets inflicted with target's Mummy
     new Ability(AbilityId.EMBODY_ASPECT_TEAL, 9)
-      .attr(PostBattleInitStatStageChangeAbAttr, [Stat.SPD], 1, true)
+      .attr(PostTeraFormChangeStatChangeAbAttr, [Stat.SPD], 1)
```
According to [smogon](https://www.smogon.com/dex/sv/abilities/embody-aspect-hearthflame/), this also needs to be a post-summon effect.

innerthunder commented on src/phases/terastallization-phase.ts:
```diff
@@ -0,0 +1,49 @@
+  public override start(): void {
+    super.start();
+
+    globalScene.phaseManager.queueMessagePhase(
+      i18next.t("battle:pokemonTerastallized", {
+        pokemonNameWithAffix: getPokemonNameWithAffix(this.pokemon),
+        type: i18next.t(`pokemonInfo:Type.${ElementalType[this.pokemon.teraType]}`),
+      }),
+    );
+
+    new CommonBattleAnim(CommonAnim.TERASTALLIZE, this.pokemon).play(false, () => {
+      this.end();
+    });
```
nit: Since the message runs after the animation, it may be less confusing if the phase is queued within the animation callback

innerthunder reviewed (commented):
code review only

DayKev committed:
Move `normalTypeMoveConversionCondition` to `ability-utils.ts`

Queue Terastallization message from animation callback
in `TerastallizationPhase`

DayKev committed:
Embody Aspect also triggers on summon

PigeonBar commented on index.html:
```diff
@@ -127,7 +127,7 @@
+          <div id="apadCycleTera" class="apad-button apad-square apad-small" data-key="CYCLE_TERA">
             <span class="apad-label">V</span>
           </div>
           <!-- buttons to display battle-specific information -->
```
The button below also needs to be updated from `"V"` to `"CYCLE_TERA"`.

PigeonBar commented on src/ui/handlers/command-ui-handler.ts:
```diff
@@ -126,23 +152,28 @@ export class CommandUiHandler extends UiHandler {
+          if ([BattleCommand.BALL, BattleCommand.RUN].includes(cursor)) {
             success = this.setCursor(cursor - 1);
+          } else if ([BattleCommand.FIGHT, BattleCommand.POKEMON].includes(cursor) && this.canTera()) {
+            success = this.setCursor(BattleCommand.TERA);
```
```suggestion
            success = this.setCursor(BattleCommand.TERA);
            this.toggleTeraButton();
```
(Should the UI also display a cursor next to the Tera button, to show more clearly that the player is selecting the Tera button?)

DayKev replied:
The Tera button is supposed to have the Terastallization shader applied to it while the cursor is on it (which this should fix, thanks).

DayKev committed:
Fix `index.html` and Tera button missing shader

PigeonBar commented on test/moves/tera-blast.test.ts:
```diff
@@ -86,34 +89,30 @@ describe("Moves - Tera Blast", () => {
-  }, 20000);
+    expect(player.getStatStage(Stat.SPATK)).toBe(-1);
+    expect(player.getStatStage(Stat.ATK)).toBe(-1);
+  });
```
```suggestion
  });

  it("should be affected by type-changing abilities (e.g., Aerilate) if user is not Terastallized", async () => {
    game.override.enemySpecies(SpeciesId.FURRET).ability(AbilityId.AERILATE);
    await game.classicMode.startBattle();

    const player = game.field.getPlayerPokemon();
    vi.spyOn(player, "getMoveType");

    game.move.use(MoveId.TERA_BLAST);
    game.setTurnOrder([BattlerIndex.PLAYER, BattlerIndex.ENEMY]);
    await game.phaseInterceptor.to("MoveEffectPhase");

    expect(player.getMoveType).toHaveLastReturnedWith(ElementalType.FLYING);
  });

  it("should NOT be affected by type-changing abilities (e.g., Aerilate) if user is Terastallized", async () => {
    game.override.enemySpecies(SpeciesId.CHIKORITA).ability(AbilityId.AERILATE);
    await game.classicMode.startBattle();
    
    const player = game.field.getPlayerPokemon();
    game.field.forceTera(player, ElementalType.NORMAL);
    vi.spyOn(player, "getMoveType");

    game.move.use(MoveId.TERA_BLAST);
    game.setTurnOrder([BattlerIndex.PLAYER, BattlerIndex.ENEMY]);
    await game.phaseInterceptor.to("MoveEffectPhase");

    expect(player.getMoveType).toHaveLastReturnedWith(ElementalType.NORMAL);
  });
```
(Some tests for #570)

PigeonBar commented on test/moves/tera-starstorm.test.ts:
```diff
@@ -29,28 +29,30 @@ describe("Moves - Tera Starstorm", () => {
-    expect(terapagos.isTerastallized()).toBe(true);
+    expect(terapagos.isTerastallized).toBe(true);
     expect(terapagos.getMoveType).toHaveReturnedWith(ElementalType.STELLAR);
   });
```
```suggestion
  });

  it("should be affected by type-changing abilities (e.g., Aerilate) if user is not Terastallized", async () => {
    game.override.ability(AbilityId.AERILATE);
    game.override.battleType("single");
    await game.classicMode.startBattle([SpeciesId.TERAPAGOS]);

    const player = game.field.getPlayerPokemon();
    vi.spyOn(player, "getMoveType");

    game.move.use(MoveId.TERA_STARSTORM);
    game.setTurnOrder([BattlerIndex.PLAYER, BattlerIndex.ENEMY]);
    await game.phaseInterceptor.to("MoveEffectPhase");

    expect(player.getMoveType).toHaveLastReturnedWith(ElementalType.FLYING);
  });
```
(A test for #570.)

PigeonBar commented on src/data/init/init-abilities.ts:
```diff
@@ -1622,29 +1621,29 @@ export function initAbilities() {
     new Ability(AbilityId.EMBODY_ASPECT_TEAL, 9)
-      .attr(PostBattleInitStatStageChangeAbAttr, [Stat.SPD], 1, true)
+      .attr(PostTeraFormChangeStatChangeAbAttr, [Stat.SPD], 1)
+      .attr(PostSummonStatStageChangeAbAttr, [Stat.SPD], 1)
```
```suggestion
      .attr(PostSummonStatStageChangeAbAttr, [Stat.SPD], 1, true)
```
(Sets `selfTarget` to `true`)

PigeonBar commented on src/data/init/init-abilities.ts:
```diff
@@ -1622,29 +1621,29 @@ export function initAbilities() {
     new Ability(AbilityId.EMBODY_ASPECT_WELLSPRING, 9)
-      .attr(PostBattleInitStatStageChangeAbAttr, [Stat.SPDEF], 1, true)
+      .attr(PostTeraFormChangeStatChangeAbAttr, [Stat.SPDEF], 1)
+      .attr(PostSummonStatStageChangeAbAttr, [Stat.SPDEF], 1)
```
```suggestion
      .attr(PostSummonStatStageChangeAbAttr, [Stat.SPDEF], 1, true)
```

PigeonBar commented on src/data/init/init-abilities.ts:
```diff
@@ -1622,29 +1621,29 @@ export function initAbilities() {
     new Ability(AbilityId.EMBODY_ASPECT_HEARTHFLAME, 9)
-      .attr(PostBattleInitStatStageChangeAbAttr, [Stat.ATK], 1, true)
+      .attr(PostTeraFormChangeStatChangeAbAttr, [Stat.ATK], 1)
+      .attr(PostSummonStatStageChangeAbAttr, [Stat.ATK], 1)
```
```suggestion
      .attr(PostSummonStatStageChangeAbAttr, [Stat.ATK], 1, true)
```

PigeonBar commented on src/data/init/init-abilities.ts:
```diff
@@ -1622,29 +1621,29 @@ export function initAbilities() {
     new Ability(AbilityId.EMBODY_ASPECT_CORNERSTONE, 9)
-      .attr(PostBattleInitStatStageChangeAbAttr, [Stat.DEF], 1, true)
+      .attr(PostTeraFormChangeStatChangeAbAttr, [Stat.DEF], 1)
+      .attr(PostSummonStatStageChangeAbAttr, [Stat.DEF], 1)
```
```suggestion
      .attr(PostSummonStatStageChangeAbAttr, [Stat.DEF], 1, true)
```

PigeonBar commented on src/data/challenge.ts:
```diff
@@ -675,6 +675,7 @@ export class FreshStartChallenge extends Challenge {
     pokemon.variant = 0; // Not shiny
     pokemon.formIndex = 0; // Froakie should be base form
     pokemon.ivs = [15, 15, 15, 15, 15, 15]; // Default IVs of 15 for all stats (Updated to 15 from 10 in 1.2.0)
+    pokemon.teraType = pokemon.species.type1; // Always primary tera type
```
Should the player also be allowed to choose the starter's secondary type as the tera type during Fresh Start, since that is also possible from a fresh save file?

DayKev replied:
Whoops, forgot that was added because of the additional restrictions in PKR. Will remove it.

PigeonBar commented on src/field/arena.ts:
```diff
@@ -39,10 +39,11 @@ export class Arena {
   public bgm: string;
   public ignoreAbilities: boolean;
   public ignoringEffectSource: BattlerIndex | null;
+  public playerTerasUsed: number = 0;
```
Attaching this data to the Arena means that the Tera Orb gets reset upon every biome change (e.g., Teleporting Hijinks, and Endless Wave X6). Is this intended? (Other game data, such as the player's amount of money, is attached to `globalScene` instead.)

DayKev replied:
Hm... for (the current state of) endless yes. Given we want to remove it though that can be ignored, so it might be better to move it to `BattleScene` for now.

PigeonBar commented on src/ui/handlers/command-ui-handler.ts:
```diff
@@ -155,6 +187,42 @@ export class CommandUiHandler extends UiHandler {
+
+    const isUltraNecrozma =
+      activePokemon.species.speciesId === SpeciesId.NECROZMA && activePokemon.getFormKey() === "ultra";
+    const isBlockedForm = activePokemon.isMega() || activePokemon.isMax() || isUltraNecrozma;
```
(TODO: Should Mega and G-Max form changes also reset the Pokemon's Tera manually?)

DayKev replied:
Yeah. Not sure where to change that though.

flx-sta replied:
Created a follow-up 
#1072

PigeonBar commented on src/ui/handlers/command-ui-handler.ts:
```diff
@@ -155,6 +187,42 @@ export class CommandUiHandler extends UiHandler {
+    const { playerTerasUsed } = globalScene.arena;
+
+    const commandPhase = this.getCommandPhase();
+    const teraCommand = globalScene.currentBattle.turnManager.findCommandFromPokemon(commandPhase.getPokemon());
```
```suggestion
    const ally = activePokemon.getAlly();
    const teraCommand = ally ? globalScene.currentBattle.turnManager.findCommandFromPokemon(ally) : undefined;
```

PigeonBar commented on src/ui/components/battle-info.ts:
```diff
@@ -588,11 +588,11 @@ export class BattleInfo extends Phaser.GameObjects.Container {
       }
 
-      const teraType = pokemon.getTeraType();
+      const teraType = pokemon.teraType;
```
```suggestion
      const teraType = pokemon.isTerastallized ? pokemon.teraType : ElementalType.UNKNOWN;
```
(Fixing a bug where Terastallizing fails to update the Tera icon, since Terastallizing no longer changes the Pokemon's `getTeraType()`.)

PigeonBar commented on src/ui/components/battle-info.ts:
```diff
@@ -336,12 +336,12 @@ export class BattleInfo extends Phaser.GameObjects.Container {
     setTextColor(this.genderText, getGenderTextStyle(pokemon.gender));
 
-    this.lastTeraType = pokemon.getTeraType();
+    this.lastTeraType = pokemon.teraType;
```
```suggestion
    this.lastTeraType = pokemon.isTerastallized ? pokemon.teraType : ElementalType.UNKNOWN;
```

PigeonBar commented on src/ui/components/battle-info.ts:
```diff
@@ -336,12 +336,12 @@ export class BattleInfo extends Phaser.GameObjects.Container {
 
     this.teraIcon.setPositionRelative(this.nameText, nameTextWidth + this.genderText.displayWidth + 1, 2);
-    this.teraIcon.setVisible(this.lastTeraType !== ElementalType.UNKNOWN);
+    this.teraIcon.setVisible(pokemon.isTerastallized);
```
```suggestion
    this.teraIcon.setVisible(pokemon.isTerastallized);
    this.teraIcon.setTintFill(Phaser.Display.Color.GetColor(...getTypeRgb(this.lastTeraType)));
```

PigeonBar reviewed (commented):
![image](https://github.com/user-attachments/assets/f0aa2aaf-2c17-49c4-90a8-a7dbbcf1d620)

DayKev committed:
Apply suggestions from review

- Add tests for Tera Blast and Tera Starstorm interacting
with move type changing abilities

- Remove forced Tera type in Fresh Start challenge

- Embody Aspect should pass `true` for `selfTarget`
in the `PostSummonStatStageChangeAbAttr` constructor

- Fix check for first pokemon queueing a Terastallization command
in double battles

Co-authored-by: PigeonBar <56974298+PigeonBar@users.noreply.github.com>

DayKev committed:
Fix Tera display in `BattleInfo` UI

Co-authored-by: PigeonBar <56974298+PigeonBar@users.noreply.github.com>

DayKev committed:
Move `playerTerasUsed` from `Arena` to `BattleScene`

PigeonBar commented on src/phases/terastallization-phase.ts:
```diff
@@ -0,0 +1,49 @@
+    this.pokemon = pokemon;
+  }
+
+  public override start(): void {
```
```suggestion
  public override start(): void {
    // A failsafe; this should no longer be possible once phase ordering for Terastallization is fixed.
    if (this.pokemon.isFainted()) {
      return super.end();
    }

```

PigeonBar commented on src/battle-scene.ts:
```diff
@@ -227,6 +228,7 @@ export default class BattleScene extends SceneBase {
   public currentBattle: Battle;
   public pokeballCounts: PokeballCounts;
   public money: number;
+  public playerTerasUsed: number = 0;
```
This also needs to be reset to 0 in `BattleScene.reset()` now.

PigeonBar commented on src/battle-scene.ts:
```diff
@@ -1372,7 +1374,18 @@ export default class BattleScene extends SceneBase {
+
+          if (
+            // In Scarlet/Violet, the player's Tera Orb automatically recharges after every battle once they've caught Terapagos
+            pokemon.species.speciesId === SpeciesId.TERAPAGOS
```
```suggestion
            (pokemon.species.speciesId === SpeciesId.TERAPAGOS && pokemon.isAllowedInChallenge())
```

DayKev committed:
Fix merge issue

PigeonBar commented on src/field/pokemon.ts:
```diff
@@ -3061,16 +3105,28 @@ export abstract class Pokemon extends Phaser.GameObjects.Container {
+      }
+
+      if (
+        source.teraType === ElementalType.STELLAR
```
```suggestion
        sourceTeraType === ElementalType.STELLAR
```

PigeonBar commented on src/phases/move-effect-phase.ts:
```diff
@@ -259,6 +259,10 @@ export class MoveEffectPhase extends HitCheckPhase {
       }
       this.updateSubstitutes();
+      const moveType = user.getMoveType(move, true);
+      if (move.isAttackMove(user, targets[0]) && !user.stellarTypesBoosted.includes(moveType)) {
```
```suggestion
      if (this.lastHit && user.isTerastallized && move.isAttackMove(user, targets[0]) && !user.stellarTypesBoosted.includes(moveType)) {
```

PigeonBar commented on src/ui/handlers/starter-select-ui-handler.ts:
```diff
@@ -3199,6 +3270,7 @@ export class StarterSelectUiHandler extends MessageUiHandler {
     this.dexAttrCursor = 0n;
```
```suggestion
    const oldTeraType = this.teraCursor !== ElementalType.UNKNOWN ? this.teraCursor : species?.type1 ?? ElementalType.UNKNOWN;
    this.dexAttrCursor = 0n;
```
(Fixing a bug where cycling nature, ability, etc. resets the selected Tera type.)

PigeonBar commented on src/ui/handlers/starter-select-ui-handler.ts:
```diff
@@ -3238,11 +3310,12 @@ export class StarterSelectUiHandler extends MessageUiHandler {
       this.abilityCursor = abilityIndex !== undefined ? abilityIndex : (abilityIndex = oldAbilityIndex);
       this.natureCursor = natureIndex !== undefined ? natureIndex : (natureIndex = oldNatureIndex);
-      const [isInParty, partyIndex]: [boolean, number] = this.isInParty(species); // we use this to firstly check if the pokemon is in the party, and if so, to get the party index in order to update the icon image
+      this.teraCursor = !isNullOrUndefined(teraType) ? teraType : (teraType = species.type1);
```
```suggestion
      this.teraCursor = !isNullOrUndefined(teraType) ? teraType : (teraType = oldTeraType);
```

PigeonBar commented on src/ui/handlers/starter-select-ui-handler.ts:
```diff
@@ -3189,7 +3260,7 @@ export class StarterSelectUiHandler extends MessageUiHandler {
   }
 
   setSpeciesDetails(species: PokemonSpecies, options: SpeciesDetails = {}): void {
```
(TODO: This function gets called with `species = null` when the cursor is on the filter bar at the top of the screen. So, the type of `species` should be `PokemonSpecies | null`.)

PigeonBar commented on src/ui/handlers/command-ui-handler.ts:
```diff
@@ -54,19 +71,24 @@ export class CommandUiHandler extends UiHandler {
     } else {
-      commandPhase = globalScene.phaseManager.getStandbyPhase() as CommandPhase;
+      this.teraButton.setVisible(false);
+      if (this.cursor === BattleCommand.TERA) {
```
```suggestion
      if (this.getCursor() === BattleCommand.TERA) {
```

PigeonBar reviewed (commented):
![image](https://github.com/user-attachments/assets/578e6b11-e77e-4f86-8c1b-456d38ec7c6c)

DayKev committed:
Reset `BattleScene#playerTerasUsed` in `BattleScene#reset`

Add check for Terapagos being allowed based on current challenges
before using it to allow infinite Tera use

DayKev committed:
Fix Stellar boost check in `MoveEffectPhase`

Co-authored-by: PigeonBar <56974298+PigeonBar@users.noreply.github.com>

DayKev committed:
Add temporary failsafe to `TerastallizationPhase` to prevent crashes

Co-authored-by: PigeonBar <56974298+PigeonBar@users.noreply.github.com>

DayKev committed:
Use `getCursor()` in `command-ui-handler.ts`

Co-authored-by: PigeonBar <56974298+PigeonBar@users.noreply.github.com>

DayKev committed:
Fix bug where cycling nature/ability/etc resets the selected Tera type

Co-authored-by: PigeonBar <56974298+PigeonBar@users.noreply.github.com>

DayKev committed:
Update STAB-related code and tests

DayKev committed:
Minor cleanup

DayKev commented:
Commits from Pigeon's review: 4ed3acf - 3291a81
New commits: 6726a8f +

flx-sta commented:
@DayKev Please resolve the conflicts

PigeonBar commented on src/phases/terastallization-phase.ts:
```diff
@@ -0,0 +1,55 @@
+  }
+
+  public override end(): void {
+    this.pokemon.isTerastallized = true;
```
```suggestion
    this.pokemon.isTerastallized = true;

    if (this.pokemon.summonData) {
      this.pokemon.summonData.addedType = null;
      this.pokemon.summonData.types = [];
    }

```

PigeonBar replied:
Also, I found the following things that modify `summonData.types` but are currently missing a check for `isTerastallized`:

- PostDefendTypeChangeAbAttr (Color Change)
- RoostedTag
- CopyBiomeTypeAttr (Camouflage)
- CopyTypeAttr (Reflect Type)
- FirstMoveTypeAttr (Conversion)
- ResistLastMoveTypeAttr (Conversion 2)
- PokemonTransformPhase

DayKev replied:
`summonData.types` shouldn't be cleared when Terastallizing because STAB is supposed to take into account the Pokémon's types at the time of Terastallization, even if they're changed.
It shouldn't be modified after a Pokémon is Terastallized though.

DayKev replied:
Replaced most of the `summonData.types = ...` with a new helper function that has a check for Terastallization in it.

DayKev committed:
Add `Pokemon#setTemporaryTypes` and update Mimicry to use it

DayKev committed:
Update remaining uses of `pokemon.summonData.types =`
